075-auto-rebalancing-cluster-scaling.md

```diff
@@ -0,0 +1,490 @@
+The `lastTransitionTime` field provides the timestamp of the latest auto-rebalancing state update.
+The `error` field provides a message if a previous auto-rebalancing failed.
+The `modes` field lists the enabled rebalancing modes, as defined through the `spec.cruiseControl.autoRebalance.mode` field, with the corresponding `brokers` list containing the brokers' IDs that are part of the scaling (up/down) actions to be useful across reconciliations.
+The brokers being added/removed are provided by the `KafkaClusterCreator` and stored within the `KafkaAssemblyOperator` for later use in the auto-rebalancing.
+
+### Auto-rebalancing execution
+
+Before describing the process in more details, it's worth mentioning how scaling up and down happens today within the Cluster Operator reconciliation flow and how the auto-rebalancing fits into it.
+
+Both the operations involve the `StrimziPodSet` controller which takes care of adding (scale up) or removing (scale down) pods by communicating with the Kubernetes API.
+
+The scaling up is triggered within the `KafkaReconciler.reconcile()` method by the `podSet()` call which updates the involved `StrimziPodSet`(s) so that the corresponding controller will take care of spinning up the newly added brokers.
+The scaling operation ends during the same reconciliation, because after having the `StrimziPodSet` controller asking for new pods through the Kubernetes API, the subsequent calls to `podsReady()`, `serviceEndpointReady()` and `headlessServiceEndpointReady()` are waiting for the newly added pods, together with the corresponding services, to be ready.
+It means that the auto-rebalancing can start in the same reconciliation, being sure that the newly added pods are available for the upcoming rebalancing operation.
+Of course, the rebalancing will take time going through several reconciliations.
+
+For the scaling down, if the brokers to be removed are hosting topics' partitions, the operation is reverted before calling the `KafkaReconciler.reconcile()` method which is called with the original replicas count as if the cluster was never scaled down.
+If the brokers to be removed are not hosting any topics' partitions, the `KafkaReconciler.reconcile()` method will be called with the new replicas count and the scaling down is triggered by the `scaleDown()` call which updates the involved `StrimziPodSet`(s) so that the corresponding controller will take care of removing the brokers.
+In this case, the auto-rebalancing can start in the same reconciliation to take topics' partitions off but going through several reconciliations.
+When the auto-rebalancing ends, the triggered pods deletion takes some time but it doesn't have any impact on the rebalancing already done.
+Of course, if the scale down check doesn't fail, it means that no topics' partitions are hosted on the brokers to be removed and no auto-rebalancing is needed.
```

Comment:
This seems badly ordered?

```suggestion
For the scaling down, if the brokers to be removed are hosting topics' partitions, the operation is reverted before calling the `KafkaReconciler.reconcile()` method which is called with the original replicas count as if the cluster was never scaled down.
In this case, the auto-rebalancing can start in the same reconciliation to take topics' partitions off but going through several reconciliations.
When the auto-rebalancing ends, the triggered pods deletion takes some time but it doesn't have any impact on the rebalancing already done.

If the brokers to be removed are not hosting any topics' partitions, the `KafkaReconciler.reconcile()` method will be called with the new replicas count and the scaling down is triggered by the `scaleDown()` call which updates the involved `StrimziPodSet`(s) so that the corresponding controller will take care of removing the brokers.
Of course, if the scale down check doesn't fail, it means that no topics' partitions are hosted on the brokers to be removed and no auto-rebalancing is needed.
```

Reply:
good catch!